src/main/java/org/jenkinsci/main/modules/sshd/AsynchronousCommand.java

```diff
@@ -95,46 +96,54 @@ public Environment getEnvironment() {
         return environment;
     }
 
+    public void start(ChannelSession channel, Environment env) throws IOException {
+        start(env);
+    }
+
     public void start(Environment env) throws IOException {
         this.environment = env;
-        thread = new Thread(new Runnable() {
-            public void run() {
-                try {
-                    int i;
-
-                    // run the command in the context of the authenticated user
-                    SecurityContext old = SecurityContextHolder.getContext();
-                    User user = getCurrentUser();
-                    if (user!=null)
-                        ACL.impersonate(user.impersonate());
-
-                    try {
-                        i = AsynchronousCommand.this.run();
-                    } finally {
-                        out.flush(); // working around SSHD-154
-                        err.flush();
-                        SecurityContextHolder.setContext(old);
-                    }
-                    callback.onExit(i);
-                } catch (Exception e) {
-                    // report the cause of the death to the client
-                    //TODO: Consider switching to UTF-8
-                    PrintWriter ps = new PrintWriter(new OutputStreamWriter(err, Charset.defaultCharset()));
-                    e.printStackTrace(ps);
-                    ps.flush();
-
-                    callback.onExit(255,e.getMessage());
-                }
-            }
-        });
+        thread = new Thread(this);
         thread.setName("SSH command: " + cmdLine.getSingleLine());
         thread.start();
     }
 
-    protected abstract int run() throws Exception;
+    protected abstract int runCommand() throws Exception;
+
+    public void run() {
+        try {
+            int i;
+
+            // run the command in the context of the authenticated user
+            SecurityContext old = SecurityContextHolder.getContext();
+            User user = getCurrentUser();
+            if (user!=null)
+                ACL.impersonate(user.impersonate());
+
+            try {
+                i = AsynchronousCommand.this.runCommand();
+            } finally {
+                out.flush(); // working around SSHD-154
+                err.flush();
+                SecurityContextHolder.setContext(old);
```

Comment:
So long as you are touching this code anyway, I would recommend switching to `ACL.as` (when `user != null`).

Reply:
done, I have changed it to use the Spring classes too.